NitroxModel/DataStructures/Optional.cs

```diff
@@ -137,16 +169,20 @@ public struct OptionalEmpty
 
     public static class Optional
     {
+        internal static Dictionary<Type, Func<object, bool>> ValueConditions = new Dictionary<Type, Func<object, bool>>();
         public static OptionalEmpty Empty { get; } = new OptionalEmpty();
 
-        public static Optional<T> Of<T>(T value)
-        {
-            return Optional<T>.Of(value);
-        }
+        public static Optional<T> Of<T>(T value) => Optional<T>.Of(value);
+        public static Optional<T> OfNullable<T>(T value) => Optional<T>.OfNullable(value);
 
-        public static Optional<T> OfNullable<T>(T value)
+        /// <summary>
+        ///     Adds a condition to the optional of the given type that is checked whenever <see cref="Optional{T}.HasValue" /> is checked.
+        /// </summary>
+        /// <param name="hasValueCondition">Condition to add to the <see cref="Optional{T}.HasValue"/> check.</param>
+        /// <param arg="T">Type that should have the extra condition. The given type will also apply to more specific types than itself.</param>
+        public static void ApplyHasValueCondition<T>(Func<T, bool> hasValueCondition) where T : class
         {
-            return Optional<T>.OfNullable(value);
+            ValueConditions.Add(typeof(T), o => hasValueCondition(o as T));
```

Comment:
Is it possible to move the type looping into this method?  Or is it done at runtime due to later dll loading?

Reply:
It's done lazily so that any more specific types (say `Object` and then `GameObject`) can be added before the checks are set in 'stone'. Doing it here would mean that other calls to `ValueConditions.Add` need to invalidate the checks cache for all optionals everytime or I need to add a "add multiple" method and do it immediatly after that call.